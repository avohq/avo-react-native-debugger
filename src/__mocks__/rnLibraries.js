jest.mock('react-native/Libraries/Settings/Settings.ios.js', () => ({
  get: jest.fn(() => null),
  set: jest.fn()
}));

jest.mock('react-native/Libraries/Utilities/BackHandler.ios.js', () => ({
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
}));
