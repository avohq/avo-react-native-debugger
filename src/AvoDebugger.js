import React from 'react';
import {Animated, TouchableOpacity} from 'react-native';
import {Component} from 'react';
import RootSiblings from 'react-native-root-siblings';
import EventsListScreen from './eventslistscreen/EventsListScreen';
import AvoBubble from './avobubble/AvoBubble';
import {styles} from './AvoDebuggerStyles';
import AvoDebuggerDrags from './AvoDebuggerDrags';
import AvoBar from './avobar/AvoBar';
import {eventsHaveErrors} from './utils';

export default class AvoDebugger extends Component {
  static avo = null;
  static rootSibling = null;
  static items = [];
  static unhandledNewItems = { count: 0 };

  static enable = (type) => {
    AvoDebugger.disable();

    let isBar = type === 'bar';

    AvoDebugger.rootSibling = new RootSiblings(<AvoDebugger
      isBar={isBar}
      ref={(avoView) => AvoDebugger.avo = avoView}
    />);
  }

  static disable = () => {
    if (AvoDebugger.rootSibling != null) {
      AvoDebugger.rootSibling.destroy();
      AvoDebugger.rootSibling = null;
      AvoDebugger.avo = null;
    }
  }

  static isEnabled = () => {
    return AvoDebugger.rootSibling !== null;
  };

  static postEvent = (
    eventId,
    timestamp,
    eventName,
    messages,
    eventProperties,
    userProperties
  ) => {
    if (AvoDebugger.isEnabled()) {
      AvoDebugger.items.push({
        key: Math.random().toString(),
        id: eventId,
        timestamp: timestamp,
        name: eventName,
        messages: messages,
        eventProps: eventProperties,
        userProps: userProperties
      });
      if (AvoDebugger.avo != null) {
        AvoDebugger.avo.onNewEvent();
      } else {
        AvoDebugger.unhandledNewItems.count += 1;
      }
    }
  };

  state = {unreadMesages: 0, pan: new Animated.ValueXY()};
  drags = new AvoDebuggerDrags();

  constructor(props) {
    super(props);
  }

  componentWillMount() {
    if (this.props.isBar) {
      this.drags.setupBarDrags(this.state.pan);
    } else {
      this.drags.setupBubbleDrags(this.state.pan);
    }
  }

  componentDidMount() {
    this.setState(prevState => ({unreadMesages: AvoDebugger.unhandledNewItems.count}), () => {
      AvoDebugger.unhandledNewItems.count = 0;
    });
  }

  onNewEvent() {
    this.setState(prevState => ({unreadMesages: prevState.unreadMesages + 1}));
  }

  render() {
    if (this.props.isBar) {
      return this.barView();
    } else {
      return this.bubbleView();
    }
  }

  barView() {
    return (
      <Animated.View
        style={[
          {
            transform: this.state.pan.getTranslateTransform()
          },
          styles.barContainer
        ]}
        {...this.drags.panResponder.panHandlers}
      >
        <TouchableOpacity
          onPress={() => {
            EventsListScreen.toggleDebuggerLogScreen(
              AvoDebugger.items,
              this.state.unreadMesages
            );
            this.setState(() => ({unreadMesages: 0}));
          }}
        >
          <AvoBar
            lastItemName={this.lastItemName()}
            lastItemTimestamp={this.lastItemTimestamp()}
            hasErrors={this.hasNewErrors()}
          />
        </TouchableOpacity>
      </Animated.View>
    );
  }

  lastItemName() {
    let lastItem = AvoDebugger.items[AvoDebugger.items.length - 1];
    if (lastItem) {
      return lastItem.name;
    } else {
      return 'No Events Yet';
    }
  }

  lastItemTimestamp() {
    let lastItem = AvoDebugger.items[AvoDebugger.items.length - 1];
    if (lastItem) {
      return lastItem.timestamp;
    } else {
      return Date.now();
    }
  }

  bubbleView() {
    return (
      <Animated.View
        style={[
          {
            transform: this.state.pan.getTranslateTransform()
          },
          styles.bubbleContainer
        ]}
        {...this.drags.panResponder.panHandlers}
      >
        <TouchableOpacity
          onPress={() => {
            EventsListScreen.toggleDebuggerLogScreen(
              AvoDebugger.items,
              this.state.unreadMesages
            );
            this.setState(() => ({unreadMesages: 0}));
          }}
        >
          <AvoBubble
            newItems={this.state.unreadMesages}
            hasErrors={this.hasNewErrors()}
          />
        </TouchableOpacity>
      </Animated.View>
    );
  }

  hasNewErrors() {
    let start = AvoDebugger.items.length - this.state.unreadMesages;
    let end = AvoDebugger.items.length;
    let newItems = AvoDebugger.items.slice(start, end);
    return eventsHaveErrors(newItems);
  }
}
