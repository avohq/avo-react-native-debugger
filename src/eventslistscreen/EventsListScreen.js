import React from 'react';
import {
  View,
  FlatList,
  BackHandler,
  Image,
  TouchableOpacity
} from 'react-native';
import {Component} from 'react';
import RootSiblings from 'react-native-root-siblings';
import PropTypes from 'prop-types';
import EventsListItem from './EventListItem';
import {images} from '../images';
import {topOffset, eventsHaveErrors} from '../utils';
import {styles} from './EventsListScreenStyles';
import {baseStyles} from '../BaseStyles';
import AvoDebugger from '../AvoDebugger';

export default class EventsListScreen extends Component {
  static logScreen = null;

  static isVisible = () => {
    return EventsListScreen.logScreen != null;
  }

  static updateDebuggerLogScreen = () => {
    if (EventsListScreen.logScreen) {
      EventsListScreen.logScreen.destroy();
      EventsListScreen.logScreen = new RootSiblings(
        <EventsListScreen items={AvoDebugger.items} />
      );
    }
  }

  static toggleDebuggerLogScreen = () => {
    if (EventsListScreen.logScreen == null) {
      EventsListScreen.logScreen = new RootSiblings(
        <EventsListScreen items={AvoDebugger.items} />
      );
    } else {
      EventsListScreen.logScreen.destroy();
      EventsListScreen.logScreen = null;
    }
  };

  componentWillMount() {
    BackHandler.addEventListener(
      'hardwareBackPress',
      this.handleBackButtonClick
    );
  }

  componentWillUnmount() {
    BackHandler.removeEventListener(
      'hardwareBackPress',
      this.handleBackButtonClick
    );
  }

  handleBackButtonClick() {
    EventsListScreen.toggleDebuggerLogScreen();
    return true;
  }

  listRow = ({item}) => (
    <EventsListItem
      item={item}
      expended={this.props.items[0] === item || eventsHaveErrors([item])}
    />
  );

  render() {
    return (
      <View style={[styles.screenContainer, {marginTop: topOffset()}]}>
        <View style={styles.avoLabelContainer}>
          <Image source={images.avo_logo} />
          <TouchableOpacity
            style={styles.closeButtonTouchableArea}
            onPress={() => {
              EventsListScreen.toggleDebuggerLogScreen();
            }}
          >
            <Image style={styles.closeButton} source={images.close_button} />
          </TouchableOpacity>
        </View>
        <View style={baseStyles.divider} />

        <FlatList
          style={styles.list}
          data={this.props.items.sort((a, b) => { return b.timestamp - a.timestamp })}
          renderItem={this.listRow}
        />
      </View>
    );
  }
}

EventsListScreen.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.exact({
      key: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      timestamp: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      messages: PropTypes.arrayOf(
        PropTypes.exact({
          tag: PropTypes.string,
          propertyId: PropTypes.string.isRequired,
          message: PropTypes.string.isRequired,
          allowedTypes: PropTypes.arrayOf(PropTypes.string),
          providedType: PropTypes.string
        })
      ),
      eventProps: PropTypes.arrayOf(
        PropTypes.exact({
          id: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          value: PropTypes.any
        })
      ),
      userProps: PropTypes.arrayOf(
        PropTypes.exact({
          id: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          value: PropTypes.any
        })
      )
    })
  )
};
