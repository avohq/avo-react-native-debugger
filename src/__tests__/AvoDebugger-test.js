import React from "react";
import * as ReactNative from "react-native";
import renderer from "react-test-renderer";

import "../__mocks__/rnLibraries";

import EventsListScreen from "../eventslistscreen/EventsListScreen";
import AvoDebugger from "../AvoDebugger";

import { avoAPI, debuggerMode, testEventItem } from "./testUtils";

global.fetch = jest.fn(() => Promise.resolve());
global.Math.random = jest.fn(() => 0.5);

const instalationID = "instalation-id-xxx";

describe("Avo Debugger", () => {
  beforeAll(() => {
    AvoDebugger.generateInstallationId = jest.fn(() => instalationID);
    AvoDebugger.init();
  });

  test("Instalation ID generated and set on debugger init", () => {
    expect(ReactNative.Settings.get).toBeCalledWith("avo_debugger_device_id");
    expect(ReactNative.Settings.set).toBeCalledWith({
      avo_debugger_device_id: instalationID
    });

    expect(AvoDebugger.installationId).toEqual(instalationID);
  });

  test("Renders correctly in Bar mode", () => {
    const avoDebugger = AvoDebugger.showDebugger({ mode: debuggerMode.BAR });
    const tree = renderer.create(avoDebugger).toJSON();

    expect(tree).toMatchSnapshot();

    expect(AvoDebugger.isEnabled()).toEqual(true);
  });

  test("Renders correctly in Bubble mode", () => {
    const avoDebugger = AvoDebugger.showDebugger({ mode: debuggerMode.BUBBLE });
    const tree = renderer.create(avoDebugger).toJSON();

    expect(tree).toMatchSnapshot();

    expect(AvoDebugger.isEnabled()).toEqual(true);
  });

  test("callTrackDebuggerStarted() sends POST request with event details", () => {
    const { name, eventProps } = testEventItem;
    fetch.mockReset();

    AvoDebugger.callTrackDebuggerStarted(name, eventProps);

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(avoAPI.TRACK, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        eventName: name,
        deviceId: instalationID,
        eventProperties: eventProps
      })
    });
  });

  test("post() updates EventsListScreen when event list is vissible", () => {
    const { timestamp, name, eventProps, messages } = testEventItem;

    EventsListScreen.isVisible = jest.fn(() => true);
    const updateLogScreenSpy = jest.spyOn(
      EventsListScreen,
      "updateDebuggerLogScreen"
    );

    AvoDebugger.avo = AvoDebugger;

    AvoDebugger.post(timestamp, name, eventProps, messages);

    expect(EventsListScreen.items.length).toEqual(1);
    expect(EventsListScreen.items).toEqual([testEventItem]);

    expect(updateLogScreenSpy).toHaveBeenCalled();
  });

  test("post() updates unreadMessage count when event list not vissible", () => {
    const { timestamp, name, eventProps, messages } = testEventItem;

    EventsListScreen.isVisible = jest.fn(() => false);

    const component = renderer.create(<AvoDebugger isBar={true} />);
    const instance = component.getInstance();

    AvoDebugger.avo = instance;

    expect(instance.state.unreadMessages).toEqual(0);

    AvoDebugger.post(timestamp, name, eventProps, messages);

    expect(instance.state.unreadMessages).toEqual(1);
  });

  test("post() updates unhandledNewItem count if debugger not initialized", () => {
    const { timestamp, name, eventProps, messages } = testEventItem;

    AvoDebugger.hideDebugger();

    AvoDebugger.post(timestamp, name, eventProps, messages);

    expect(AvoDebugger.unhandledNewItems.count).toEqual(1);
  });

  test("hideDebugger() disables debugger", () => {
    AvoDebugger.showDebugger({ mode: debuggerMode.BAR });

    const destroySpy = jest.spyOn(AvoDebugger.rootSibling, "destroy");

    AvoDebugger.hideDebugger();

    expect(destroySpy).toHaveBeenCalled();
    expect(AvoDebugger.rootSibling).toBeNull();
    expect(AvoDebugger.avo).toBeNull();

    expect(AvoDebugger.isEnabled()).toEqual(false);
  });

  test("lastItemName() and lastItemTimestamp() returns last event data for AvoBar", () => {
    const { name, eventProps, messages } = testEventItem;

    const component = renderer.create(<AvoDebugger isBar={true} />);
    const instance = component.getInstance();

    AvoDebugger.post(+new Date(2020, 9, 15, 10, 5, 1), `${name}-0`, eventProps, messages);
    AvoDebugger.post(+new Date(2020, 9, 15, 10, 5, 2), `${name}-1`, eventProps, messages);
    AvoDebugger.post(+new Date(2020, 9, 15, 10, 5, 3), `${name}-2`, eventProps, messages);

    const lastItem = instance.lastItemName();
    const lastTimestamp = instance.lastItemTimestamp();

    expect(lastItem).toEqual(`${name}-2`);
    expect(lastTimestamp).toEqual(+new Date(2020, 9, 15, 10, 5, 3));
  });


  test("hasNewErrors() validates new events for errors", () => {
    const {timestamp, name, eventProps, messages } = testEventItem;

    const component = renderer.create(<AvoDebugger isBar={true} />);
    const instance = component.getInstance();
    AvoDebugger.avo = instance;

    EventsListScreen.items = [];
    instance.state.unreadMessages = 0;
  
    EventsListScreen.isVisible = jest.fn(() => true);
    AvoDebugger.post(+new Date(2020, 9, 15, 10, 5, 3), `${name}-visible`, eventProps, []);
    
    expect(instance.hasNewErrors()).toEqual(false);
    
    EventsListScreen.isVisible = jest.fn(() => false);
    AvoDebugger.post(+new Date(2020, 9, 15, 10, 5, 4), `${name}-no-error`, eventProps, []);
    
    expect(instance.hasNewErrors()).toEqual(false);

    AvoDebugger.post(+new Date(2020, 9, 15, 10, 5, 5), `${name}-error`, eventProps, messages);
    AvoDebugger.post(+new Date(2020, 9, 15, 10, 5, 6), `${name}-no-error`, eventProps, []);

    expect(instance.hasNewErrors()).toEqual(true);
  });
});
