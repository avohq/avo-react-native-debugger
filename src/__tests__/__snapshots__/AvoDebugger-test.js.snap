// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avo Debugger Renders correctly in Bar mode 1`] = `null`;

exports[`Avo Debugger Renders correctly in Bubble mode 1`] = `null`;
